File: rust/bron_kerbosch/src/bron_kerbosch2.rs
//! Bron-Kerbosch algorithm with pivot picked arbitrarily

use crate::bron_kerbosch_pivot::{visit, PivotChoice};
use crate::graph::{connected_vertices, UndirectedGraph, VertexSetLike};
use crate::reporter::Reporter;

pub fn explore<Graph, Rprtr>(graph: &Graph, reporter: &mut Rprtr)
where
    Graph: UndirectedGraph,
    Rprtr: Reporter,
{
    let candidates = connected_vertices(graph);
    if !candidates.is_empty() {
        visit(
            graph,
            reporter,
            PivotChoice::Arbitrary,
            PivotChoice::Arbitrary,
            candidates,
            Graph::VertexSet::new(),
            None,
        );
    }
}

File: go/src/bron_kerbosch/bron_kerbosch2.go
package bron_kerbosch

func bron_kerbosch2(graph *UndirectedGraph) [][]Vertex {
	// Bron-Kerbosch algorithm with pivot of highest degree
	candidates := graph.connected_vertices()
	if candidates.IsEmpty() {
		return nil
	}
	excluded := make(VertexSet, len(candidates))
	var reporter SimpleReporter
	visit_max_degree(graph, &reporter, candidates, excluded, nil)
	return reporter.cliques
}
